### ISSUE TITLE
Strange Y axis values with statsd counters

### ISSUE THREAD

**Issue by gtaylor** [graphite-project/graphite-web#704]
I've got a statsd counter used to record each time certain events happen. It seems like statsd and ultimately graphite are receiving the messages just fine, but I'm getting weird fractional Y axis value:

http://i.imgur.com/hCCDU5A.png

I'm always incrementing by 1, so I'm expecting the Y axis to be 0 to some arbitrary integer instead of 0 to arbitrary fractional values.

Any ideas? This seems like super weird behavior.

Here's my storage-schemas.conf:

```
# Schema definitions for Whisper files. Entries are scanned in order,
# and first match wins. This file is scanned for changes every 60 seconds.
#
#  [name]
#  pattern = regex
#  retentions = timePerPoint:timeToStore, timePerPoint:timeToStore, ...

# Carbon's internal metrics. This entry should match what is specified in
# CARBON_METRIC_PREFIX and CARBON_METRIC_INTERVAL settings
[carbon]
pattern = ^carbon\.
retentions = 60:90d

[stats]
priority = 110
pattern = .*
retentions = 10:2160,60:10080,600:262974
```

**Comment by ihaworth** [comment]
I've been seeing similar issues.  I've tracked it down to being dependent upon the size of the time range over which the data is being viewed.  For example, the following is the same data set shown over 3 different size time intervals:

2 Months:
![2month](https://cloud.githubusercontent.com/assets/1285306/8779385/2572de96-2efb-11e5-9d52-dfaa535fdad1.PNG)

1 Week:
![1week](https://cloud.githubusercontent.com/assets/1285306/8779415/448962aa-2efb-11e5-870d-00acdaa7e997.PNG)

1 Day:
![1day](https://cloud.githubusercontent.com/assets/1285306/8779424/4dddefa6-2efb-11e5-805d-70484f8636c9.PNG)

The left-hand graphs are the monotonic count, and the right-hand graphs are the derivative.  As you can see, each change should be a count of 1, so all the derivatives should be 1.  However, they're around 0.04 for the 2 month graph, 0.33 for the 1 week graph, and (correctly) 1 for the 1 day graph.

This illustrates that the underlying data is sound (as confirmed with the format=raw parameter) but the rendering of it gets more inaccurate as the time interval increases.

This isn't a critical issue for me, but it would be nice to see a solution.

**Comment by sravanireddyl** [comment]
I am new to graphite graphs, I am also facing the same problem with y axis value. 
Any one got sollution?

**Comment by deniszh** [comment]
A I only one who doesn't get what the problem is?
For me all graph looks completely sane. 
As documentations of derifative() says " This function does not normalize for periods of time, as a true derivative would. Instead see the perSecond() function to calculate a rate of change over time."

**Comment by sravanireddyl** [comment]
I have a counter metric which increases by 1 at a time,
the first graph renders when I select time range of 1 day, and 2nd graph rendered when I select time range of 1 week.
The y-axis value changing according to time but i need the fixed Y-axis value without fractional values(exact count of metric).
![graph2](https://cloud.githubusercontent.com/assets/13868724/9355752/6cb703da-4699-11e5-8dbc-54831e1ea429.PNG)
![graph1](https://cloud.githubusercontent.com/assets/13868724/9355753/7070d55a-4699-11e5-9eec-f16c385e3ff3.PNG)

**Comment by deniszh** [comment]
@sravanireddyl:
What you mean "i need the fixed Y-axis value without fractional values(exact count of metric)"? 
You need logins per second/minute/hour ? Use ```perSecond()``` and scale it as documentation says.
Or, if ```perSecond()``` is not in your Graphite - use ```scaleToSeconds(nonNegativeDerivative(...),1)``` - which is exactly the same. But it still be fractional too, I think, because of rounding errors.

**Comment by deniszh** [comment]
```scaleToSeconds(nonNegativeDerivative(...),60)``` will give you logins per minute, ```scaleToSeconds(nonNegativeDerivative(...),3600)``` - per hour/etc

**Comment by sravanireddyl** [comment]
@deniszh 
Thank you for your reply.
I have applied functions "scaletoseconds" and "nonNegativeDerivative", still am getting the graph with fractional values as below.  
![graph3](https://cloud.githubusercontent.com/assets/13868724/9375845/3623e04c-4723-11e5-8946-27b313e441fb.PNG)

How can i identify number of logins per minute with that graph.?
(This question seems to be silly to you but I am very new to all these graphs and dashboards.)

**Comment by deniszh** [comment]
```scaleToSeconds(nonNegativeDerivative(...),60)``` will give you number of logins per minute. But I don't think it will be exact integer, because of nature of derivative and rounding errors.

**Comment by sravanireddyl** [comment]
But the value changing when I change the date range.

**Comment by deniszh** [comment]
Then it's depends on your storage schema. By default retention function is ```avg```, but for counters it's need to be ```sum```. Check this article for details - http://www.franklinangulo.com/blog/2014/5/20/graphite-series-3-whisper-storage-schemas-aggregations

**Comment by bilal-fazlani** [comment]
Is this issue resolved ? I am facing the same issue

**Comment by deniszh** [comment]
@bilal-fazlani - it's not Graphite issue, it's your storage schema / aggregation issue. Could you please check links above -  http://www.franklinangulo.com/blog/2014/5/20/graphite-series-3-whisper-storage-schemas-aggregations ?
What's your `/opt/graphite/conf/storage-schemas.conf` and `/opt/graphite/conf/storage-aggregation.conf` looks like?

**Comment by bilal-fazlani** [comment]
storage-aggregation.conf

```
[min]
pattern = \.lower$
xFilesFactor = 0.1
aggregationMethod = min

[max]
pattern = \.upper(_\d+)?$
xFilesFactor = 0.1
aggregationMethod = max

[sum]
pattern = \.sum$
xFilesFactor = 0
aggregationMethod = sum

[count]
pattern = \.count$
xFilesFactor = 0
aggregationMethod = sum

[count_legacy]
pattern = ^stats_counts.*
xFilesFactor = 0
aggregationMethod = sum

[default_average]
pattern = .*
xFilesFactor = 0.3
aggregationMethod = average
```

storage-schemas.conf

```
[default]
pattern = .*
retentions = 10s:6h,1min:6d,10min:365d

[catch_em_all]
pattern = .*
retentions = 10s:6h,1min:6d,10min:365d
```

**Comment by tterrace** [comment]
This confused me for a while as well. I sent a few test counts via nc and the Y values on the graph are incorrect for a 60 minute time range, but correct for 59 minutes or below. 

Here's at a 60 time range:

![image](https://cloud.githubusercontent.com/assets/6719894/24890774/bb2fc82e-1e3f-11e7-810c-216a081f0c61.png)

Then at 59 minutes:

![image](https://cloud.githubusercontent.com/assets/6719894/24890784/c908a04c-1e3f-11e7-83fe-0686a10bf427.png)

You can see the Y values in the second screen shot are correct, but the first screen shot shows values that are half of what they should be.

**Comment by loisaidasam** [comment]
For anyone still struggling with this, it seems that the fine folks over at Etsy have gone into a bit of detail writing about it and how to fix it:

https://github.com/etsy/statsd/blob/master/docs/graphite.md

# TL;DR:

The default [aggregationMethod](http://graphite.readthedocs.io/en/latest/config-carbon.html#storage-aggregation-conf) for storage is `average` (example config: [storage-aggregation.conf.example](https://github.com/tmm1/graphite/blob/master/carbon/conf/storage-aggregation.conf.example)), where as for counters it should be `sum`.

# To fix:

Either

## 1. Counter end with `.count`

Make your counter key end with `.count`, for example `my.data.point.count`. This will hit the `[sum]` config and use `aggregationMethod = sum`

Or

## 2. [statsd] in `storage_aggregation.conf`

Add the following section to `/opt/graphite/conf/storage-aggregation.conf` before the `[default_average]` section:

```
[statsd]
pattern = ^stats_counts.*
xFilesFactor = 0
aggregationMethod = sum
```
